h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java

```diff
@@ -109,7 +109,7 @@ else if (f.vec(i-catCols).isInt())
      * @param weightsVector Vector with row weights, possibly null
      * @return A sum of chunk lengths. Possibly zero, if there are no chunks or the chunks are empty.
      */
-    private static long sumChunksLength(int[] chunkIds, Vec vec, Vec weightsVector, int[] chunkLengths) {
+    public static long sumChunksLength(int[] chunkIds, Vec vec, Vec weightsVector, int[] chunkLengths) {
```

Comment:
This methods itself, originally written by me I believe, could be a major slow down, as it iterates of whole vec in a sequential way. As I was the one to introduce the problem, I offer to make it distributed @honzasterba .

Reply:
Nevermind, I forgot this is only done for local chunks. :man_facepalming: 

Could be calculated beforehand for the whole cluster anyway, if possible and not calculate it each round. Something like this:

```java
    private static class ChunksLenTask extends MRTask<ChunksLenTask> {
        public int[] chunkLengths;

        ChunksLenTask(final int nChunks) {
            chunkLengths = new int[nChunks];
            Arrays.fill(chunkLengths, -1);
        }

        @Override
        public void map(Chunk c0, Chunk weightsChunk) {
            chunkLengths[c0.cidx()] = c0._len;
            int lastNonZeroIndex = weightsChunk.nextNZ(-1, true);

            while (weightsChunk.nextNZ(lastNonZeroIndex, true) != -1) {
                chunkLengths[c0.cidx()]--;
                lastNonZeroIndex = weightsChunk.nextNZ(lastNonZeroIndex, true);
            }

        }

        @Override
        public void reduce(ChunksLenTask mrt) {
            for (int i = 0; i < mrt.chunkLengths.length; i++) {
                if (mrt.chunkLengths[i] != -1) {
                    chunkLengths[i] = mrt.chunkLengths[i];
                }
            }
        }

        public long nonZeroChunkLength() {
            return ArrayUtils.suml(chunkLengths);
        }
    }
```